File: showTeamStats.rb
require 'libRuby'
gStyle.SetOptStat(0)
gStyle.SetTitleFillColor(10)
gStyle.SetTitleBorderSize(0)
gStyle.SetPalette(1,0)

def printOptions
  puts ""
  puts "showTeamStats.rb Usage options"
  puts ""
  puts "\t-F (string) currently existing file name, or name of output ROOT file (no default)[Optional]"
  puts "\t-t (string) name of team whose stats to plot (no default)[Required if -F not used]"
  puts "\t-H calculate and plot gameScore percentages"
  puts "\t-s (int) (double) srs value to use for weights and histograms (default: \"fixed\")[Optional]"
  puts "\t-g draw game-by-game statistics [Optional]"
  puts "\t-h print this message"
  puts ""
  puts "Ex: ruby showTeamStats.rb -t \"2015 north carolina\" -H -s 0.5"
  puts
  puts "This program is used to show the results of the showTeamStats program, but"
  puts "it can also be used to run that program first and then process the output, so as to make"
  puts "this a one-step process.  If a ROOT file already exists, you can plot its graphs"
  puts "using the -F switch to set the input file name. If a ROOT file does not already exist,"
  puts "then you can create one by using the -t option to set the name of the team you want to plot."
  puts "If you wish to save the ROOT file created, simply use the"
  puts "-F option to set the name, it will be saved by default in the rootFiles/ directory, so"
  puts "only the name is necessary.  If no name is set, then a temporary file is created which"
  puts "is deleted at the end of the program."
  puts ""
end

fileName = nil
drawGames = false
teamName = nil
useHistograms = false
srsVal = "free"
#command line switcher
ARGV.each_with_index do |entry, index|
  case entry
  when /^-F/
    fileName = ARGV[index+1]
  when /^-H/
    useHistograms = true
  when /^-s/
    srsVal = ARGV[index+1]
  when /^-t/
    teamName = ARGV[index+1]
  when /^-D|-d|-g|-G/
    drawGames = true
  end
end

file = nil
if teamName.nil?
  if fileName.nil?
    puts "Must enter a teamName or a fileName"
    exit
  else
    file = TFile.new(fileName)
  end
else
  if fileName.nil?
    if !useHistograms
      `$CLION/showTeamStats -t \"#{teamName}\" -o \"temp.root\"`
    else
      `$CLION/showTeamStats -t \"#{teamName}\" -o \"temp.root\" -H -s #{srsVal}`
    end
    file = TFile.new("rootFiles/temp.root")
  else
    if !useHistograms
      `$CLION/showTeamStats -t \"#{teamName}\" -o \"#{fileName}\"`
    else
      `$CLION/showTeamStats -t \"#{teamName}\" -o \"#{fileName}\" -H -s #{srsVal}`
    end
    file = TFile.new("rootFiles/#{fileName}")
  end
end

avgs_per_time = {}
avgs_per_game = {}
wavgs_per_time = {}
wavgs_per_game = {}
league_avgs = {}
lines = {}
realLines = {}
ss = ["oor.p","oefg.p","oto.p","oftmr.p","srs",
      "dor.p","defg.p","dto.p","dftmr.p","rpi"]

ss.each do |s|
  avgs_per_time[s] = gROOT.FindObject("avg_per_time_#{s.gsub(".","")}")
  wavgs_per_time[s] = gROOT.FindObject("wavg_per_time_#{s.gsub(".","")}")
  avgs_per_game[s] = gROOT.FindObject("avg_per_game_#{s.gsub(".","")}")
  wavgs_per_game[s] = gROOT.FindObject("wavg_per_game_#{s.gsub(".","")}")
  #For some idiot reason I don't understand I can't draw a line read in this way
  #so I have to make a new TLine object using it.
  lines[s] = gROOT.FindObject("#{s.gsub(".","")}Line")
  realLines[s] = TLine.new(lines[s].GetX1,lines[s].GetY1,lines[s].GetX2,lines[s].GetY2)
end

if useHistograms
  avgs_per_time["gameScore"] = gROOT.FindObject("avg_per_time_gameScore")
  wavgs_per_time["gameScore"] = gROOT.FindObject("wavg_per_time_gameScore")
  realLines["gameScore"] = realLines["rpi"]
end

can = TCanvas.new("can","",0,0,1300,700)
can.Divide(5,2)

ss.each_with_index do |s,i|
  can.cd(i+1)
  avgs_per_time[s].SetMinimum(0)
  avgs_per_time[s].SetMaximum(1)
  avgs_per_time[s].SetMinimum(-30) if s == "srs"
  avgs_per_time[s].SetMaximum(30) if s == "srs"
  avgs_per_time[s].SetMarkerStyle(4)
  avgs_per_time[s].SetMarkerColor(1)
  avgs_per_time[s].SetLineColor(1)
  avgs_per_time[s].SetMarkerSize(0.5)
  avgs_per_time[s].Draw("psame")
  wavgs_per_time[s].SetMarkerStyle(8)
  wavgs_per_time[s].SetMarkerColor(4)
  wavgs_per_time[s].SetMarkerSize(0.5)
  wavgs_per_time[s].Draw("psame")

  if drawGames
    avgs_per_game[s].SetMarkerStyle(8)
    avgs_per_game[s].SetMarkerColor(1)
    avgs_per_game[s].SetMarkerSize(0.5)
    avgs_per_game[s].Draw("psame")
    wavgs_per_game[s].SetMarkerStyle(8)
    wavgs_per_game[s].SetMarkerColor(2)
    wavgs_per_game[s].SetMarkerSize(0.5)
    wavgs_per_game[s].Draw("psame")
  end

  realLines[s].SetLineColor(2)
  realLines[s].Draw("same")
end

if useHistograms
  can.cd(10)
  avgs_per_time["gameScore"].SetMinimum(0)
  avgs_per_time["gameScore"].SetMaximum(1)
  avgs_per_time["gameScore"].SetMarkerStyle(4)
  avgs_per_time["gameScore"].SetMarkerColor(1)
  avgs_per_time["gameScore"].SetLineColor(1)
  avgs_per_time["gameScore"].SetMarkerSize(0.5)
  avgs_per_time["gameScore"].Draw("p")
  wavgs_per_time["gameScore"].SetMarkerStyle(8)
  wavgs_per_time["gameScore"].SetMarkerColor(4)
  wavgs_per_time["gameScore"].SetMarkerSize(0.5)
  wavgs_per_time["gameScore"].Draw("psame")

  realLines["gameScore"].SetLineColor(2)
  realLines["gameScore"].Draw("same")
end

if (!teamName.nil? and fileName.nil?)
  `rm rootFiles/temp.root`
end

##This sets it so that the application ends when a canvas is closed. 
##Without it the application can hang in the background if you 
##don't exit properly.
gApplication.Connect("TCanvas","Closed()","TApplication",gApplication,
                     "Terminate()")
##If runROOT selected the window will stay until closed,  
##If not selected the window goes away after the fits run. 
 gApplication.Run# if runROOT             

File: showTeamStats.cpp
#include <iostream>
#include <fstream>
#include "TFile.h"
#include "TH1F.h"
#include "TLine.h"
#include <boost/algorithm/string.hpp>
#include <TROOT.h>
#include "src/Team.h"
#include "src/ConstantTeam5YearAverages.h"
#include "src/ConstantSeasonInfo.h"
#include "src/ConstantGameFunction.h"
#include "src/ConstantSRSadditions.h"
#include "src/ConstantTeamNeutralRatios.h"
#include "src/ConstantWAverageFunctions.h"
#include "src/ConstantStandardDeviations.h"

void printOptions();

int main(int argc,char *argv[]){
    int c;
    int year = 0;
    std::string teamName = "", outFileName = "";
    bool useHistogramsFile = false;
    std::string srsValue = "free";

    /*____________________________Parse Command Line___________________________*/
    while((c = getopt(argc,argv,"t:o:hHs:")) != -1){
        switch(c){
            case 't':
                teamName = optarg;
                break;
            case 'o':
                outFileName = optarg;
                break;
            case 's':
                srsValue.assign(optarg);
                break;
            case 'H':
                useHistogramsFile = true;
                break;
            case 'h':
                printOptions();
                return 0;
            default:
                // not an option
                break;
        }
    }

    if (teamName == "" || outFileName == ""){
        std::cout << "You must specify a team with the -t switch and an outFileName with the -o switch" << std::endl;
        printOptions();
        return 0;
    }

    year = stoi(teamName.substr(0,4));

    char *homePath, path[256], histName[256];
    homePath = getenv("HOME");

    sprintf(path, "%s/cpp/NCAA_C/constants/team_5yr_averages.d", homePath);
    ConstantTeam5YearAverages *averages = ConstantTeam5YearAverages::Instance();
    averages->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/season_info.d", homePath);
    ConstantSeasonInfo *seasonInfo = ConstantSeasonInfo::Instance();
    seasonInfo->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/waverage_functions.d", homePath);
    ConstantWAverageFunctions *functions = ConstantWAverageFunctions::Instance();
    functions->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/team_neutral_ratios.d", homePath);
    ConstantTeamNeutralRatios *ratios = ConstantTeamNeutralRatios::Instance();
    ratios->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/srs_additions.d", homePath);
    ConstantSRSadditions *additions = ConstantSRSadditions::Instance();
    additions->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/game_function_weights.%s.d", homePath,srsValue.c_str());
    ConstantGameFunction *gameFunction = ConstantGameFunction::Instance();
    gameFunction->initialize(path);

    sprintf(path, "%s/cpp/NCAA_C/constants/waverage_standard_deviations.d", homePath);
    ConstantStandardDeviations *stdDevs = ConstantStandardDeviations::Instance();
    stdDevs->initialize(path);

    //read in the averages and waverages for this one team
    Team *team = new Team(teamName);
    sprintf(path, "%s/cpp/NCAA_C/teams/%i/teams.%s.averages.d",
            homePath,team->getYear(),boost::replace_all_copy(teamName," ","_").c_str());
    team->addAverages(path);
    sprintf(path, "%s/cpp/NCAA_C/teams/%i/teams.%s.waverages.d",
            homePath,team->getYear(),boost::replace_all_copy(teamName," ","_").c_str());
    team->addWAverages(path);

    TFile *histsFile;
    std::unordered_map<int, TH1F*> probs_by_year, probs_err_by_year;
    if (useHistogramsFile){
        sprintf(path, "%s/cpp/NCAA_C/rootFiles/gameFunctionHistograms.%s.root", homePath, srsValue.c_str());
        histsFile = new TFile(path);

        for (int y = 2007; y <= 2016; y++){
            sprintf(path, "probs_by_year_%d", y);
            probs_by_year.emplace(y, (TH1F*)gROOT->FindObject(path));
            sprintf(path, "probs_err_by_year_%d", y);
            probs_err_by_year.emplace(y, (TH1F*)gROOT->FindObject(path));
        }
    }

    //create the output ROOT file
    sprintf(path, "%s/cpp/NCAA_C/rootFiles/%s",homePath,outFileName.c_str());
    TFile outFile(path,"recreate");

    //create and fill the containers for the averages and waverages
    std::unordered_map<std::string, TH1F*> avgs_per_time, wavgs_per_time, avgs_per_game, wavgs_per_game;
    std::unordered_map<std::string, TLine*> lines;
    std::string stats[] = {"oor.p", "oefg.p", "oto.p", "oftmr.p", "srs",
                           "dor.p", "defg.p", "dto.p", "dftmr.p", "rpi"};
    for (std::string &s : stats){
        sprintf(histName, "avg_per_time_%s",(boost::replace_all_copy(s,".","").c_str()));
        avgs_per_time.emplace(s, new TH1F(histName,s.c_str(),140,0,140));
        sprintf(histName, "wavg_per_time_%s",(boost::replace_all_copy(s,".","").c_str()));
        wavgs_per_time.emplace(s, new TH1F(histName,s.c_str(),140,0,140));

        sprintf(histName, "avg_per_game_%s",(boost::replace_all_copy(s,".","").c_str()));
        avgs_per_game.emplace(s, new TH1F(histName,s.c_str(),140,0,140));
        sprintf(histName, "wavg_per_game_%s",(boost::replace_all_copy(s,".","").c_str()));
        wavgs_per_game.emplace(s, new TH1F(histName,s.c_str(),140,0,140));
    }
    if (useHistogramsFile) {
        avgs_per_time.emplace("gameScore", new TH1F("avg_per_time_gameScore", "gameScore", 140, 0, 140));
        wavgs_per_time.emplace("gameScore", new TH1F("wavg_per_time_gameScore", "gameScore", 140, 0, 140));
    }

    //create the lines that hold the league averages for each stat
    lines.emplace("oor.p", new TLine(0,averages->get(year,"oor.p"),140,
                                     averages->get(year,"oor.p")));
    lines["oor.p"]->Write("oorpLine");
    lines.emplace("dor.p", new TLine(0,averages->get(year,"oor.p"),140,
                                     averages->get(year,"oor.p")));
    lines["dor.p"]->Write("dorpLine");
    lines.emplace("oefg.p", new TLine(0,averages->get(year,"oefg.p"),140,
                                      averages->get(year,"oefg.p")));
    lines["oefg.p"]->Write("oefgpLine");
    lines.emplace("defg.p", new TLine(0,averages->get(year,"oefg.p"),140,
                                      averages->get(year,"oefg.p")));
    lines["defg.p"]->Write("defgpLine");
    lines.emplace("oftmr.p", new TLine(0,averages->get(year,"oftmr.p"),140,
                                       averages->get(year,"oftmr.p")));
    lines["oftmr.p"]->Write("oftmrpLine");
    lines.emplace("dftmr.p", new TLine(0,averages->get(year,"oftmr.p"),140,
                                       averages->get(year,"oftmr.p")));
    lines["dftmr.p"]->Write("dftmrpLine");
    lines.emplace("oto.p", new TLine(0,averages->get(year,"oto.p"),140,
                                     averages->get(year,"oto.p")));
    lines["oto.p"]->Write("otopLine");
    lines.emplace("dto.p", new TLine(0,averages->get(year,"oto.p"),140,
                                     averages->get(year,"oto.p")));
    lines["dto.p"]->Write("dtopLine");
    lines.emplace("srs", new TLine(0,0,140,0));
    lines["srs"]->Write("srsLine");
    lines.emplace("rpi", new TLine(0,0.5,140,0.5));
    lines["rpi"]->Write("rpiLine");

    //bring the averages here locally
    std::unordered_map<std::string, TeamAverage *> averageHash = team->getAveragesByDate();

    //some placeholders
    TeamAverage *last_average;
    TeamWAverage *waverage, *last_waverage;

    //some calendar tools
    boost::gregorian::date_duration dateDuration(1);
    boost::gregorian::days indDuration;

    //cycle through the averages hash
    for (auto &average : averageHash){
        last_average = team->AverageOnDate(average.second->getDate() - dateDuration);
        if (!last_average) continue; //skip the first date
        waverage = team->WAverageOnDate(average.second->getDate());
        last_waverage = team->WAverageOnDate(average.second->getDate() - dateDuration);

        int last_num_games = last_average->getNum_games();
        indDuration = average.second->getDate() - seasonInfo->get(team->getYear(),"season start");
        int ind = (int)indDuration.days();

        //cycle through the strings
        for (std::string &s : stats) {
            if (boost::contains(s, ".p") == 1){
                //only do this on days where there was a game.
                if (average.second->getNum_games() != last_num_games) {
                    //to get the percentage for a game, we take the difference between
                    //the makes and attempts today and the day before, and then taking the
                    //retio.
                    double m = average.second->getValue(boost::replace_first_copy(s, ".p", ".m")) -
                               last_average->getValue(boost::replace_first_copy(s, ".p", ".m"));
                    double a = average.second->getValue(boost::replace_first_copy(s, ".p", ".a")) -
                               last_average->getValue(boost::replace_first_copy(s, ".p", ".a"));
                    double p = m / a;
                    avgs_per_game[s]->SetBinContent(ind,p);

                    m = waverage->getValue(boost::replace_first_copy(s, ".p", ".m")) -
                        last_waverage->getValue(boost::replace_first_copy(s, ".p", ".m"));
                    a = waverage->getValue(boost::replace_first_copy(s, ".p", ".a")) -
                        last_waverage->getValue(boost::replace_first_copy(s, ".p", ".a"));
                    p = m / a;
                    wavgs_per_game[s]->SetBinContent(ind,p);
                }
                avgs_per_time[s]->SetBinContent(ind,average.second->getValue(s));
                wavgs_per_time[s]->SetBinContent(ind,waverage->getValue(s));
            }
            else if (boost::contains(s, ".m") == 1){
                //makes have to take into account the fact that I store the total number of makes
                //(or total number of weighted attempts) for each day, not the average.

                if (average.second->getNum_games() != last_num_games) {
                    avgs_per_game[s]->SetBinContent(ind,average.second->getValue(s) -
                            last_average->getValue(s));
                    wavgs_per_game[s]->SetBinContent(ind,waverage->getValue(s) -
                            last_waverage->getValue(s));
                }
                avgs_per_time[s]->SetBinContent(ind,average.second->getValue(s) / (double)average.second->getNum_games());
                wavgs_per_time[s]->SetBinContent(ind,waverage->getValue(s) / (double)waverage->getNum_games());
            }
            else if (s == "srs") {
                avgs_per_time[s]->SetBinContent(ind,waverage->getOrigSRS());
                wavgs_per_time[s]->SetBinContent(ind,waverage->getSrs());
            }
            else if (s == "rpi") {
                wavgs_per_time[s]->SetBinContent(ind,waverage->getRpi());
            }
        }
        if (useHistogramsFile) {
            double gameScore = gameFunction->predictGame(waverage, team->getYear(), "neutral", "neutral");
            double gameScorePct, gameScorePct_err;
            if (gameScore >= 3) {
                gameScorePct = 1;
                gameScorePct_err = 1;
            }
            else if (gameScore <= -3) {
                gameScorePct = 0;
                gameScorePct_err = 1;
            }
            else {
                gameScorePct = probs_by_year[team->getYear()]->GetBinContent((gameScore + 3) * 1600 / 6.0);
                gameScorePct_err = probs_err_by_year[team->getYear()]->
                        GetBinContent((gameScore + 3) * 1600 / 6.0);
            }
            avgs_per_time["gameScore"]->SetBinContent(ind, gameScorePct);
            avgs_per_time["gameScore"]->SetBinError(ind, gameScorePct_err);
            wavgs_per_time["gameScore"]->SetBinContent(ind, gameScorePct);
            wavgs_per_time["gameScore"]->SetBinError(ind, gameScorePct_err);
        }
    }

    outFile.Write();
    outFile.Close();

    return 0;
}

void printOptions(){
    std::cout << std::endl;
    std::cout << "showTeamStats Usage options" << std::endl;
    std::cout << "" << std::endl;
    std::cout << "\t-t (string) teamname (no default)[Required]" << std::endl;
    std::cout << "\t-o (string) outFileName (no default)[Required]" << std::endl;
    std::cout << "\t-s (double) srs value to use for weights and histograms (default: \"fixed\")[Optional]" << std::endl;
    std::cout << "\t-H calculate game score using histograms file (no default)[Optional]" << std::endl;
    std::cout << "" << std::endl;
    std::cout << "Ex: showTeamStats -t \"2015 north carolina\" -o \"2015unc.root\" -H -s 0.5" << std::endl;
    std::cout << std::endl;
    std::cout << "This program creates histograms for each of the four factors, offense and" << std::endl;
    std::cout << "defense, over the course of the season.  It creates histograms for both" << std::endl;
    std::cout << "the regular averages as well as the weighted averages.  It also plots" << std::endl;
    std::cout << "SRS and gameScore or RPI.  You view the resultant ROOT histograms" << std::endl;
    std::cout << "using the showTeamStats.rb Ruby program.  All you have to specify is the" << std::endl;
    std::cout << "team name, and the name of the output file.  You can also decide to " << std::endl;
    std::cout << "plot the game score by selecting the -H option, and specifying an" << std::endl;
    std::cout << "SRS value for the weights if necessary." << std::endl;
    std::cout << std::endl;
}